Promises api

edulan opened the pull request:
# What
Use ES6 Promises instead of callbacks so we can easily chain actions and handle errors.

Usage:
```javascript
var autoLaunch = new AutoLaunch({name: 'AwesomeApp'});

autoLaunch.enable().then(function() {
  // Do stuff when enabled
});

autoLaunch.disable().then(function() {
  // Do stuff when disabled
});

autoLaunch.isEnabled().then(function(enabled) {
  if (enabled) {
    // Do something if enabled
  } else {
    // Do something if disabled
  }
});

// Detect when something is not working
autoLaunch.disable().catch(function(error) {
  // Handle error
});
```

# Concerns
Only compatible with node versions >= 11.13.

andresgutgon commented:
:green_apple: :+1:

4ver commented on src/AutoLaunchLinux.coffee:
```diff
@@ -10,30 +10,39 @@ module.exports =
+
+            mkdirp.sync(@getDir())
+            fs.writeFile file, data, (err) ->
+                return reject if err?
```
Would be best to pass the error when calling reject.

4ver commented on src/AutoLaunchLinux.coffee:
```diff
@@ -10,30 +10,39 @@ module.exports =
+            file = @getFile(opts)
+
+            fs.stat file, (err, stat) ->
+                resolve(stat?)
```
```coffee
return reject err if err?
```

edulan replied:
My concern here is how we differentiate between disable and error states. For instance in Linux implementation a `ENOENT` error means disable state, while `EACCES` indicates a real error.

The same applies for Windows but is even more difficult to tackle due to the lack of information provided by `winreg` lib. They've added support for capture error messages in [new versions](https://github.com/fresc81/node-winreg/pull/23), maybe we could update to latest version in order to get that info.

In Mac error and returned data are separated so there's nothing to change.

So I'm wondering how can we approach this... Should we differentiate error types? What are your thoughts?

4ver replied:
I guess, considering we don't currently handle these types of errors well and of the amount of work involved; we can let it go in this pr. Maybe add a `# TODO: Error handling`. I'll add a separate issue to get this fixed. Would be good to be able to return consistent error messages regardless of the platform (when we can).

Thanks for the detailed response!

4ver commented:
Am all for this :+1:  If you could get the tests working again that would be great.

Thanks for your contribution! :dancer:

edulan commented:
@4ver I've added TODO reminders and fixed some error stuff for Linux adapter. I agree with you to separate error handling to another PR :+1:.

Also keep in mind Promises support is for node versions >= 11.13. Don't know if you want to create a special branch/version due to incompatibility issues with previous releases. Other option would be to add a polyfill package, but IMHO I'd keep the library as simple and dependency-free as possible.

Thank you for your consideration :smile:

4ver commented:
Sorry for the delay in getting back to you. I think we should go ahead and just use https://github.com/jakearchibald/es6-promise. It's pretty lightweight so shouldn't be much of an issue.

Thanks @edulan

edulan commented:
Cool! I'll include es6-promise then :+1:
